File: find_contours.py
import cv2
import numpy as np


class FindContours:
    def __init__(self, grid):
        self.count = 0
        self.contours = {}
        self.LNBD = 1
        self.NBD = 1
        self.Disp_with_number = True
        self.grid = grid.copy().astype("int32")
        self.grid = np.pad(self.grid, ((1, 1), (1, 1)), 'constant', constant_values=0)  # add edges
        self.MAX_BODER_NUMBER = self.grid.shape[0] * self.grid.shape[1]
        self.contours_dict = {1: self.Contour(-1, "Hole")}

    def Contour(self, parent, contour_type, start_point=[-1, -1]):
        contour = {"parent": parent,
                   "contour_type": contour_type,
                   "son": [],
                   "start_point": start_point}  # Hole/Outer
        return contour

    def find_neighbor(self, center, start, clock_wise=1):
        weight = -1
        if clock_wise == 1:
            weight = 1
        neighbors = np.array([[0, 0], [0, 1], [0, 2], [1, 2], [2, 2], [2, 1], [2, 0], [1, 0]])
        indexs = np.array([[0, 1, 2],
                           [7, 9, 3],
                           [6, 5, 4]])
        start_ind = indexs[start[0] - center[0] + 1][start[1] - center[1] + 1]
        for i in range(1, len(neighbors) + 1):
            cur_ind = (start_ind + i * weight + 8) % 8
            x = neighbors[cur_ind][0] + center[0] - 1
            y = neighbors[cur_ind][1] + center[1] - 1
            if self.grid[x][y] != 0:
                return [x, y]
        return [-1, -1]

    def board_follow(self, center_p, start_p, mode):
        ij = center_p
        ij2 = start_p
        ij1 = self.find_neighbor(ij, ij2, 1)
        x = ij1[0]
        y = ij1[1]
        if ij1 == [-1, -1]:
            self.grid[ij[0]][ij[1]] = -self.NBD
            return
        ij2 = ij1
        ij3 = ij
        self.count += 1
        self.contours[self.count] = []
        for k in range(self.MAX_BODER_NUMBER):
            ij4 = self.find_neighbor(ij3, ij2, 0)
            self.contours[self.count].append([ij4[0]-1, ij4[1]-1])
            x = ij3[0]
            y = ij3[1]
            if ij4[0] - ij2[0] <= 0:
                weight = -1
            else:
                weight = 1
            if self.grid[x][y] < 0:
                self.grid[x][y] = self.grid[x][y]

            elif self.grid[x][y - 1] == 0 and self.grid[x][y + 1] == 0:
                self.grid[x][y] = self.NBD * weight

            elif self.grid[x][y + 1] == 0:
                self.grid[x][y] = -self.NBD

            elif self.grid[x][y] == 1 and self.grid[x][y + 1] != 0:
                self.grid[x][y] = self.NBD

            else:
                self.grid[x][y] = self.grid[x][y]

            if ij4 == ij and ij3 == ij1:
                return
            ij2 = ij3
            ij3 = ij4

    def raster_scan(self):
        for i in range(self.grid.shape[0]):
            self.LNBD = 1
            for j in range(self.grid.shape[1]):
                if abs(self.grid[i][j]) > 1:
                    self.LNBD = abs(self.grid[i][j])
                if self.grid[i][j] >= 1:
                    if self.grid[i][j] == 1 and self.grid[i][j - 1] == 0:
                        self.NBD += 1
                        self.board_follow([i, j], [i, j - 1], 1)
                        border_type = "Outer"

                    elif self.grid[i][j] >= 1 and self.grid[i][j + 1] == 0:
                        border_type = "Hole"
                        self.NBD += 1
                        self.board_follow([i, j], [i, j + 1], 1)
                    else:
                        continue
                    parent = self.LNBD
                    if self.contours_dict[self.LNBD]["contour_type"] == border_type:
                        parent = self.contours_dict[self.LNBD]["parent"]
                    self.contours_dict[self.NBD] = self.Contour(parent, border_type, [i - 1, j - 1])
                    self.contours_dict[parent]["son"].append(self.NBD)
        self.grid = self.grid[1:-1, 1:-1]  # cut edge


def expand_img_point(img, expand):
    new_img = np.ones((img.shape[0] * expand, img.shape[1] * expand, 3)) * 255
    for i in range(img.shape[0]):
        for j in range(img.shape[1]):
            if img[i][j] == 1:
                new_img[i*expand][j*expand][:] = 0
    return new_img


def main():
    img = cv2.imread("map2.png")
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    ret, binary = cv2.threshold(src=gray, thresh=50, maxval=1, type=cv2.THRESH_BINARY)
    fc = FindContours(binary)
    fc.raster_scan()
    print(fc.contours_dict)
    print(fc.contours)
    # set contours green
    for i in fc.contours:
        for point in fc.contours[i]:
            img[point[0], point[1]] = (0, 255, 0)
    cv2.imwrite("./res/res.png", img)
    cv2.imshow("img", img)
    cv2.waitKey(0)
    # get a point map
    expand = 5
    point_map = expand_img_point(binary, expand)
    cv2.imwrite("./res/point_map_origin.png", point_map)
    cv2.imshow("point_map_origin", point_map)
    cv2.waitKey(0)
    # draw contours on point map
    for i in fc.contours:
        for id in range(len(fc.contours[i])):
            cv2.line(point_map, (expand * fc.contours[i][id][1], expand * fc.contours[i][id][0]), (
            expand * fc.contours[i][(id + 1) % len(fc.contours[i])][1],
            expand * fc.contours[i][(id + 1) % len(fc.contours[i])][0]), (0, 0, 255), 1)
    cv2.imwrite("./res/point_map.png", point_map)
    cv2.imshow("point_map", point_map)
    cv2.waitKey(0)

if __name__ == "__main__":
    main()

File: FindContours.java
import java.util.*;

class FindContours {
    private int count = 0;
    public Map<Integer, ArrayList<int[]>> contours = new HashMap<>(); // 存放每一个轮廓
    private int LNDB = 1;
    private int NBD = 1;
    private int[][] grid;
    private int MAX_BODER_NUMBER = 0;
    public Map<Integer, Map<String, Object>> contours_dict = new HashMap<>();

    // 构造函数
    public FindContours(ArrayList<ArrayList<Integer>> agrid) {
        this.grid = pad_grid(agrid); // 加边
        this.MAX_BODER_NUMBER = this.grid.length * this.grid[0].length;
        this.contours_dict.put(1, this.Contour(-1, "Hole", new int[0]));
    }

    public Map<String, Object> Contour(int parent, String contour_type, int[] start_point) {
        if (start_point.length == 0) start_point = new int[]{-1, -1};
        return new HashMap<String, Object>(Map.of("parent", parent, "contour_type", contour_type, "son", new ArrayList<Integer>(), "start_point", start_point));
    }

    public static int[][] pad_grid(ArrayList<ArrayList<Integer>> agrid) {
        agrid.add(0, new ArrayList<Integer>(Collections.nCopies(agrid.get(0).size(),0)));
        agrid.add(new ArrayList<Integer>(Collections.nCopies(agrid.get(0).size(),0)));

        for (ArrayList<Integer> i : agrid) {
            i.add(0);
            i.add(0, 0);
        }
        //ArrayList<ArrayList<Integer>>转int[][]
        int[][] res = agrid.stream().map(i -> i.stream().mapToInt(j -> j).toArray()).toArray(int[][]::new);
        return res;
    }

    public int[] find_neighbor(int[] center, int[] start, boolean clock_wise) {
        int weight = -1;
        if (clock_wise) weight = 1;
        int[][] neighbors = new int[][]{{0, 0}, {0, 1}, {0, 2}, {1, 2}, {2, 2}, {2, 1}, {2, 0}, {1, 0}};
        int[][] indexs = new int[][]{{0, 1, 2},
                {7, 9, 3},
                {6, 5, 4}};
        int start_ind = indexs[start[0] - center[0] + 1][start[1] - center[1] + 1];
        for (int i = 1; i < neighbors.length + 1; i++) {
            int cur_ind = (start_ind + i * weight + 8) % 8;
            int x = neighbors[cur_ind][0] + center[0] - 1;
            int y = neighbors[cur_ind][1] + center[1] - 1;
            if (this.grid[x][y] != 0)
                return new int[]{x, y};
        }
        return new int[]{-1, -1};
    }

    public void board_follow(int[] center_p, int[] start_p) {
        int[] ij = center_p;
        int[] ij2 = start_p;
        int[] ij1 = this.find_neighbor(ij, ij2, true);
        if (Arrays.equals(ij1, new int[]{-1, -1})) {
            this.grid[ij[0]][ij[1]] = -this.NBD;
            return;
        }
        ij2 = ij1;
        int[] ij3 = ij;
        this.count += 1;
        this.contours.put(this.count, new ArrayList<>());
        for (int k = 0; k < this.MAX_BODER_NUMBER; k++) {
            int[] ij4 = this.find_neighbor(ij3, ij2, false);
            this.contours.get(this.count).add(new int[]{ij4[0] - 1, ij4[1] - 1});
            int x = ij3[0];
            int y = ij3[1];
            int weight = 0;
            if ((ij4[0] - ij2[0]) <= 0) weight = -1;
            else weight = 1;
            if (this.grid[x][y] < 0)
                this.grid[x][y] = this.grid[x][y];

            else if (this.grid[x][y - 1] == 0 && this.grid[x][y + 1] == 0)
                this.grid[x][y] = this.NBD * weight;

            else if (this.grid[x][y + 1] == 0)
                this.grid[x][y] = -this.NBD;

            else if (this.grid[x][y] == 1 && this.grid[x][y + 1] != 0)
                this.grid[x][y] = this.NBD;

            else
                this.grid[x][y] = this.grid[x][y];

            if (Arrays.equals(ij4, ij) && Arrays.equals(ij3, ij1))
                return;

            ij2 = ij3;
            ij3 = ij4;
        }
    }

    public void raster_scan() {
        for (int i = 0; i < this.grid.length; i++) {
            this.LNDB = 1;
            for (int j = 0; j < this.grid[0].length; j++) {
                if (Math.abs(this.grid[i][j]) > 1)
                    this.LNDB = Math.abs(this.grid[i][j]);
                if (this.grid[i][j] >= 1) {
                    String border_type;
                    if (this.grid[i][j] == 1 && this.grid[i][j - 1] == 0) {
                        this.NBD += 1;
                        this.board_follow(new int[]{i, j}, new int[]{i, j - 1});
                        border_type = "Outer";
                    } else if (this.grid[i][j] >= 1 && this.grid[i][j + 1] == 0) {
                        border_type = "Hole";
                        this.NBD += 1;
                        this.board_follow(new int[]{i, j}, new int[]{i, j + 1});
                    } else
                        continue;
                    int parent = this.LNDB;
                    if (this.contours_dict.get(this.LNDB).get("contour_type").equals(border_type))
                        parent = (int) this.contours_dict.get(this.LNDB).get("parent");
                    this.contours_dict.put(this.NBD, this.Contour(parent, border_type, new int[]{i - 1, j - 1}));
                    ((ArrayList<Integer>) this.contours_dict.get(parent).get("son")).add(this.NBD);
                }

            }
        }
        // TODO: cut edges

    }

    public static ArrayList<ArrayList<Integer>> readMap(String path) {
        ArrayList<ArrayList<Integer>> a = new ArrayList<>();
        try {
            java.io.File file = new java.io.File(path);
            java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.FileReader(file));
            String line = null;
            while ((line = reader.readLine()) != null) {
                ArrayList<Integer> b = new ArrayList<>();
                for (int i = 0; i < line.length(); i++) {
                    int temp = line.charAt(i) - '0';
                    if (temp == 0 || temp == 1)
                        b.add(temp);
                }
                a.add(b);
            }
            reader.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return a;
    }

    public static void main(String[] args) {
        ArrayList<ArrayList<Integer>> aa = FindContours.readMap("map0.txt");
//        ArrayList<ArrayList<Integer>> aa = FindContours.readMap("map1.txt");
        FindContours map = new FindContours(aa);
        map.raster_scan();
        // 每一个轮廓的顺序就存储在map.contours里
        System.out.println("contours num: " + map.contours.size());

        for (Map.Entry<Integer, ArrayList<int[]>> entry : map.contours.entrySet()) {
            System.out.println("contour" + entry.getKey() + ", points: ");
            for (int[] i : entry.getValue()) {
                System.out.print(Arrays.toString(i));
            }
            System.out.println();
        }
    }

}






























